storybookjs/storybook
# UI: Add configurable tags-based exclusion from sidebar/autodocs

Closes N/A

## What I did

- [x] Added a new parameter, `docs.stories.filter` to control which stories show up in the stories block
- [x] Added a new `main.ts` field called `tags` to configure tag behavior
	- `tags['tag-name'].excludeFromSidebar` to hide stories from the sidebar
	- `tags['tag-name'].excludeFromDocsStories` to hide stories from the Stories block
- [x] Added several built-in tags:
	- `'dev-only': { excludeFromDocsStories: true }`
	- `'docs-only': { excludeFromSidebar: true }`
	- `'test-only': { excludeFromSidebar: true, excludeFromDocsStories: true }`

The default `docs.stories.filter` is backwards compatible with the previous behavior which filters out stories with a truthy `docs.disable` parameter.

## Checklist for Contributors

### Testing

Added e2e tests in the telescoping PR https://github.com/storybookjs/storybook/pull/25537

#### The changes in this PR are covered in the following automated tests:

#### Manual testing

1. Open a sandbox
2. Add `tags: ['dev-only']` to a story & observe it disappear from autodocs
2. Add `tags: ['docs-only']` to a story & observe it disappear from the sidebar
3. Add `tags: ['test-only']` to a story & observe it disappear from both the sidebar and autodocs
4. Add a new tag `foo` to the main.ts configuration & restart storybook:

```js
// .storybook/main.js
export default {
  // stories, etc.
  tags: {
    foo: {  excludeFromSidebar: true,  excludeFromDocsStories: true },
  },
};
```

### Documentation

👉 Needs docs



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release


This pull request has been released as version [`0.0.0-pr-25328-sha-5fe992e9`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25328-sha-5fe992e9). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-25328-sha-5fe992e9`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25328-sha-5fe992e9) |
| **Triggered by** | @shilman |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`shilman/autodocs-filter`](https://github.com/storybookjs/storybook/tree/shilman/autodocs-filter) |
| **Commit** | [`5fe992e9`](https://github.com/storybookjs/storybook/commit/5fe992e9201fece88571cf8e8a918a0113ff5faf) |
| **Datetime** | Tue Jan  2 10:32:24 UTC 2024 (`1704191544`) |
| **Workflow run** | [7384765903](https://github.com/storybookjs/storybook/actions/runs/7384765903) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=25328`_
</details>
